feat(pos): add camera scanning functionality for barcode/qr code

engahmed1190 opened the pull request:
- Add new CameraScanner.vue component with scanning capabilities
- Integrate scanner with ItemsSelector.vue for barcode search
- Add POS Profile fields to enable/configure camera scanning
- Include html5-qrcode library as dependency
- Implement scanning feedback and error handling
#61 

This feature still beta needed to be tested and verified in multiple camera formate and multiple QR and Scan Codes

engahmed1190 committed:
feat(pos): add camera scanning functionality for barcode/qr code

- Add new CameraScanner.vue component with scanning capabilities
- Integrate scanner with ItemsSelector.vue for barcode search
- Add POS Profile fields to enable/configure camera scanning
- Include html5-qrcode library as dependency
- Implement scanning feedback and error handling

Manaa0-0 commented:
@engahmed1190 does this needs SSL certificate and domain for the browser to allow permission denied in http sites

engahmed1190 commented:
may be http doesn't allow camera. You may try to set a production site for Frappe. I will try to see if we can have work around. 

@Manaa0-0 you enabled it from POS Profile. I added a setting check box to enable the camera search.

engahmed1190 committed:
feat(scanner): replace html5-qrcode with qr-scanner and quagga for auto-detection

- Replace html5-qrcode library with qr-scanner and quagga for unified QR/barcode scanning
- Implement auto-detection of code formats with parallel scanning capabilities
- Add scan format detection and display in UI
- Improve scanner UI with persistent dialog and better visual feedback
- Update package.json dependencies and add required types

engahmed1190 commented:
Now it's better scanning. I used a better library for scanning, and it works over HTTP. 

![image](https://github.com/user-attachments/assets/4652f097-aa2c-4709-a734-be3fd4b5297a)

engahmed1190 committed:
refactor(Navbar): improve socket connection handling with environment detection

- Add environment detection (development, production, fallback) to determine appropriate socket URL
- Simplify and organize socket connection logic with helper functions
- Maintain same connection behavior while making code more maintainable

Manaa0-0 commented:
@engahmed1190 that amazing 👏

engahmed1190 commented:
Let's test it very well. @Manaa0-0

Manaa0-0 commented:
@engahmed1190 I will suit it up now also please @defendicon could you review it As I see working on the same feauters

engahmed1190 commented:
Sorry , I haven't seen that @defendicon . I thought you were working on Offline Mode. Please check this and let me know if you have any comments. Thanks

Manaa0-0 commented:
@engahmed1190 I try but it  doesn't  reach to camera ether in phone nor cumputer
![Screenshot from 2025-05-31 23-16-56](https://github.com/user-attachments/assets/21586b83-3dc2-4fba-ba20-60d0f4c2a768)

Manaa0-0 commented:
@engahmed1190 I use HTTPS in another one and can reach the camera but the problem is can't read the bare code while when using the default one it work fine and identified the product my suggestion is to use the default and put it in POS awesome

defendicon commented:
@engahmed1190 @Manaa0-0 i will look tomorrow. Great work 👍 👏 I was tired on working offline project thats why i take a little break and start work on this feature.

defendicon commented:
I have Checked and implemented new solution which created on new branch Cam_PR and add 

Scanning Speed Enhanced almost 500 times.

Multiple Code Types: It can detect not only QR codes but also other 1D barcode formats (like EAN, UPC, Code 128, etc.). For this, the formats prop needs to be used.

Unified Solution: In our current setup, we're using qr-scanner for QR codes and QuaggaJS for 1D barcodes. vue-qrcode-reader can replace both, simplifying our code and making maintenance easier. This library uses the BarcodeDetector API (if available in the browser) or zxing-wasm (a WebAssembly port of the ZXing C++ library), which is a powerful and well-regarded barcode processing engine.

defendicon commented:
@engahmed1190 
now I will tweaks some bugs. please check this branch code and pull request again with this code or if you find any bugs please check or fix.

defendicon commented:
@engahmed1190 I am stuck in Cam_PR branch of package.json file how can I add code which runs "yarn install" first so it install newly added dependencies and then build process start so it don't give error. We can run separately yarn install but non technical users cant run or update directly with bench build command.

engahmed1190 commented:
```
{
  "scripts": {
    // Starts the Vite development server for local development
    "dev": "vite",

    // 💡 Pre-Build Hook:
    // Runs automatically before the "build" script.
    // This ensures that any new dependencies added to package.json are installed.
    // Prevents build errors caused by missing modules.
    "prebuild": "yarn install",

    // 🔨 Build Process:
    // Builds the project using esbuild with production mode and the specified apps.
    // The "prebuild" hook will run first to make sure dependencies are up-to-date.
    "build": "node esbuild --production --apps posawesome --run-build-command"
  }
}
```

✅ **How it works:**
- `yarn build` triggers the `prebuild` script (`yarn install`) automatically.
- After dependencies are installed, the build process (`node esbuild ...`) executes.

defendicon commented:
@engahmed1190 Ok i will add tomorrow if you can add this code please add and make pull request so i can merge it. Did you check Cam_PR branch ?

engahmed1190 commented:
Done

defendicon commented:
@engahmed1190 I got error i think it should be npm install rather than yarn install ?

pleskinovox@189:~/frappe-bench/apps/posawesome$ bench build --app posawesome
✔ Application Assets Linked

yarn run v1.22.22
$ node esbuild --production --apps posawesome --run-build-command
✘ [ERROR] Could not resolve "vue-qrcode-reader"

sfc-script:/home/pleskinovox/frappe-bench/apps/posawesome/posawesome/public/js/posapp/components/pos/CameraScanner.vue?type=script:2:29:
  2 │ import { QrcodeStream } from 'vue-qrcode-reader';
    ╵                              ~~~~~~~~~~~~~~~~~~~
You can mark the path "vue-qrcode-reader" as external to exclude it from the bundle, which will
remove this error.

✘ [ERROR] [plugin frappe-vue-style] Cannot read properties of undefined (reading 'outputs')

/home/pleskinovox/frappe-bench/apps/frappe/esbuild/frappe-vue-style.js:10:41:
  10 │       let files = get_files(result.metafile.outputs);
     ╵                                             ^

at /home/pleskinovox/frappe-bench/apps/frappe/esbuild/frappe-vue-style.js:10:42
at /home/pleskinovox/frappe-bench/apps/frappe/node_modules/esbuild/lib/main.js:1052:21
at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
This error came from the "onEnd" callback registered here:

/home/pleskinovox/frappe-bench/apps/frappe/esbuild/frappe-vue-style.js:9:8:
  9 │     build.onEnd((result) => {
    ╵           ~~~~~

at setup (/home/pleskinovox/frappe-bench/apps/frappe/esbuild/frappe-vue-style.js:9:9)
at handlePlugins (/home/pleskinovox/frappe-bench/apps/frappe/node_modules/esbuild/lib/main.js:855:23)
ERROR There were some problems during build

Error: Build failed with 2 errors:
sfc-script:/home/pleskinovox/frappe-bench/apps/posawesome/posawesome/public/js/posapp/components/pos/CameraScanner.vue?type=script:2:29: ERROR: Could not resolve "vue-qrcode-reader"
/home/pleskinovox/frappe-bench/apps/frappe/esbuild/frappe-vue-style.js:10:41: ERROR: [plugin: frappe-vue-style] Cannot read properties of undefined (reading 'outputs')
at failureErrorWithLog (/home/pleskinovox/frappe-bench/apps/frappe/node_modules/esbuild/lib/main.js:1624:15)
at /home/pleskinovox/frappe-bench/apps/frappe/node_modules/esbuild/lib/main.js:1266:28
at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
Terminated
error Command failed with exit code 143.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Traceback (most recent call last):
File "", line 198, in _run_module_as_main
File "", line 88, in _run_code
File "/home/pleskinovox/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 114, in
main()
File "/home/pleskinovox/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 20, in main
click.Group(commands=commands)(prog_name="bench")
File "/home/pleskinovox/frappe-bench/env/lib/python3.12/site-packages/click/core.py", line 1161, in call
return self.main(*args, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/home/pleskinovox/frappe-bench/env/lib/python3.12/site-packages/click/core.py", line 1082, in main
rv = self.invoke(ctx)
^^^^^^^^^^^^^^^^
File "/home/pleskinovox/frappe-bench/env/lib/python3.12/site-packages/click/core.py", line 1697, in invoke
return _process_result(sub_ctx.command.invoke(sub_ctx))
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/home/pleskinovox/frappe-bench/env/lib/python3.12/site-packages/click/core.py", line 1697, in invoke
return _process_result(sub_ctx.command.invoke(sub_ctx))
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/home/pleskinovox/frappe-bench/env/lib/python3.12/site-packages/click/core.py", line 1443, in invoke
return ctx.invoke(self.callback, **ctx.params)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/home/pleskinovox/frappe-bench/env/lib/python3.12/site-packages/click/core.py", line 788, in invoke
return _callback(*args, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/home/pleskinovox/frappe-bench/apps/frappe/frappe/commands/utils.py", line 77, in build
bundle(
File "/home/pleskinovox/frappe-bench/apps/frappe/frappe/build.py", line 255, in bundle
frappe.commands.popen(command, cwd=frappe_app_path, env=get_node_env(), raise_err=True)
File "/home/pleskinovox/frappe-bench/apps/frappe/frappe/commands/init.py", line 97, in popen
raise subprocess.CalledProcessError(return, command)
subprocess.CalledProcessError: Command 'yarn run production --apps posawesome --run-build-command' returned non-zero exit status 143.

engahmed1190 commented:
It works with me when I do

- `yarn build`
- `bench build --app posawesome`

defendicon commented:
@engahmed1190 
How to intimate people about the update process of posawesome app because everyone will run
git pull
bench build --app posawesome ( But Got Error Because yarn build is not run first or they dont know that yarn build should be run first )

engahmed1190 commented:
```
 "scripts": {
    "dev": "vite",
    "prebuild": "npm install"
  }
```

 removed the build command then run `bench build --app posawesome` , it builds in one step . test it at your side
